TulparAudio/internal/include/tulpar/internal/BufferCollection.hpp

```diff
@@ -0,0 +1,167 @@
+/*
+* Copyright (C) 2018 by Godlike
+* This code is licensed under the MIT license (MIT)
+* (http://opensource.org/licenses/MIT)
+*/
+
+#ifndef TULPAR_INTERNAL_BUFFER_COLLECTION_HPP
+#define TULPAR_INTERNAL_BUFFER_COLLECTION_HPP
+
+#include <tulpar/internal/Collection.hpp>
+
+#include <tulpar/audio/Buffer.hpp>
+
+#include <tulpar/TulparAudio.hpp>
+
+#include <mule/asset/Handler.hpp>
+
+#include <chrono>
+#include <cstdint>
+#include <string>
+#include <unordered_map>
+
+namespace tulpar
+{
+namespace internal
+{
+
+/** @brief  Handles basic buffer id operations */
+struct OpenAVBufferHandler
+{
+    /** @brief  Generates a collection of buffer handles via OpenAL
+     *
+     *  Underlying buffer objects registered with OpenAL are tied to device
+     *  object that is active at the moment of calling this method
+     *
+     *  @param  batchSize   number of handles to generate
+     *
+     *  @return a collection of newly generated buffer handles
+     */
+    static Collection<audio::Buffer>::Handles Generate(uint32_t batchSize);
+
+    /** @brief  Resets given handle
+     *
+     *  @note   Does nothing at the moment
+     *
+     *  @param  handle  valid buffer handle
+     */
+    static void Reclaim(Collection<audio::Buffer>::Handle handle);
+
+    /** @brief  Returns given collection of handles to OpenAL
+     *
+     *  @param  handle  collection of valid buffer handles
+     */
+    static void Delete(Collection<audio::Buffer>::Handles const& handles);
+};
+
+/** @brief  Collection working with audio buffers */
+class BufferCollection
+    : public Collection<audio::Buffer>
+{
+public:
+    //! Shortcut to buffer handle type
+    using Handle = audio::Buffer::Handle;
+
+    //! Shortcut to buffer migration mapping as hashmap<old, new>
+    using MigrationMapping = std::unordered_map<audio::Buffer::Handle, audio::Buffer::Handle>;
+
+    /** @brief  Constructs buffer collection object
+     *
+     *  @param  generator   functor that will be called when generating
+     *                      new handle
+     *  @param  reclaimer   functor that will be called when reclaiming
+     *                      previously used handle
+     *  @param  deleter     functor that will be called when releasing
+     *                      previously used handle
+     */
+    BufferCollection(
+        Collection<audio::Buffer>::HandleGenerator generator = OpenAVBufferHandler::Generate
```

Comment:
Should a user code be able to provide a custom generator without providing custom reclaimer and or deleter? Would it make sense to either provide all use default for everything?

Reply:
This is exposed mainly for testing. But yes, if a user was to extend some functionality around default `OpenAVBufferHandler::` methods, it would make sense only to overload any combination of the three.
Most likely to be overloaded is reclaimer. If user caches buffer handles on their side, they would write a reclaimer that would invalidate their cached stuff and call `OpenAVBufferHandler::Reclaim` when they're done.

Overloading generator and deleter is done in tests.